File: python/euler008.py
"""
A solution to Project Euler Problem 8.

https://projecteuler.net/problem=8
"""

import functools
import itertools
import operator
from typing import Generator


# From itertools examples
# https://docs.python.org/2.3/lib/itertools-example.html
def window(seq, n=2):
    """
    Returns a sliding window (of width n) over data from the iterable.

    s -> (s0,s1,...s[n-1]), (s1,s2,...,sn), ...
    """
    it = iter(seq)
    result = tuple(itertools.islice(it, n))
    if len(result) == n:
        yield result
    for elem in it:
        result = result[1:] + (elem,)
        yield result


def digit_window_products(digits: str, n: int) -> Generator[int, None, None]:
    """
    Returns the product of the digits in a sliding window of width n
    over the string of digits.

    >>> gen = digit_window_products('1234', 2)
    >>> next(gen)
    2
    >>> next(gen)
    6
    >>> next(gen)
    12
    >>> next(gen)
    Traceback (most recent call last):
    ...
    StopIteration
    """
    for win in window(digits, n):
        yield functools.reduce(operator.mul, map(int, win))


def solution(digits: str, n: int) -> int:
    """
    Returns the greatest product of n adjacent digits in the given
    digits string.

    >>> solution(INPUT, 4)
    5832
    """
    return max(digit_window_products(digits, n))


if __name__ == '__main__':
    with open("../resources/input008.txt", "r") as fin:
        digits = fin.read().replace("\n", "")

    print(solution(digits, 13))

File: haskell/Euler008.hs
module Euler008 where

import Data.Char (isDigit, digitToInt)
import Data.Maybe (mapMaybe)
import Data.List (tails)

-- | The window size given in the problem.
windowSize :: Int
windowSize = 13

-- | Converts a character to the integer it represents if possible.
charToInt :: Char -> Maybe Int
charToInt c
  | isDigit c = Just $ digitToInt c
  | otherwise = Nothing

-- | Yields a sliding window of the given length over the given list.
--
-- The last n elements of the produced list will contain fewer than n elements,
-- where n is the window size.
--
-- @
-- windows 2 [1..4] == [[1,2],[2,3],[3,4],[4],[]]
-- @
windows :: Int -> [a] -> [[a]]
windows n = map (take n) . tails

-- | Computes the greatest product of a series of n adjacent elements in the
-- given list of integers.
solution :: Int -> [Int] -> Int
solution n = maximum .  map product . windows n

main :: IO ()
main = do
        input <- readFile "../resources/input008.txt"
        let digits = mapMaybe charToInt input
        print $ solution windowSize digits
